File: app/src/main/java/dev/ragnarok/fenrir/model/Conversation.java
package dev.ragnarok.fenrir.model;

import static dev.ragnarok.fenrir.util.Utils.firstNonEmptyString;

import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.Utils;

public class Conversation {

    private final int id;

    private String title;

    private int unreadCount;

    private String photo50;

    private String photo100;

    private String photo200;

    /**
     * ID of the last read incoming message.
     */
    private int inRead;

    /**
     * ID of the last read outcoming message.
     */
    private int outRead;

    private Owner interlocutor;

    private Keyboard currentKeyboard;

    private Message pinned;

    private boolean isGroupChannel;

    private int acl;

    private int major_id;

    private int minor_id;

    public Conversation(int id) {
        this.id = id;
    }

    public int getAcl() {
        return acl;
    }

    public Conversation setAcl(int acl) {
        this.acl = acl;
        return this;
    }

    public Message getPinned() {
        return pinned;
    }

    public Conversation setPinned(Message pinned) {
        this.pinned = pinned;
        return this;
    }

    public Owner getInterlocutor() {
        return interlocutor;
    }

    public Conversation setInterlocutor(Owner interlocutor) {
        this.interlocutor = interlocutor;
        return this;
    }

    public int getId() {
        return id;
    }

    public String getTitle() {
        if (Peer.isUser(id)) {
            String custom = Settings.get().other().getUserNameChanges(id);
            if (!Utils.isEmpty(custom)) {
                return custom;
            }
        }
        return title;
    }

    public Conversation setTitle(String title) {
        this.title = title;
        return this;
    }

    public Keyboard getCurrentKeyboard() {
        return currentKeyboard;
    }

    public Conversation setCurrentKeyboard(Keyboard currentKeyboard) {
        this.currentKeyboard = currentKeyboard;
        return this;
    }

    public int getUnreadCount() {
        return unreadCount;
    }

    public Conversation setUnreadCount(int unreadCount) {
        this.unreadCount = unreadCount;
        return this;
    }

    public String getPhoto50() {
        return photo50;
    }

    public Conversation setPhoto50(String photo50) {
        this.photo50 = photo50;
        return this;
    }

    public String getPhoto100() {
        return photo100;
    }

    public Conversation setPhoto100(String photo100) {
        this.photo100 = photo100;
        return this;
    }

    public String getPhoto200() {
        return photo200;
    }

    public Conversation setPhoto200(String photo200) {
        this.photo200 = photo200;
        return this;
    }

    public int getInRead() {
        return inRead;
    }

    public Conversation setInRead(int inRead) {
        this.inRead = inRead;
        return this;
    }

    public int getOutRead() {
        return outRead;
    }

    public Conversation setOutRead(int outRead) {
        this.outRead = outRead;
        return this;
    }

    public String get100orSmallerAvatar() {
        return firstNonEmptyString(photo100, photo50);
    }

    public String getMaxSquareAvatar() {
        return firstNonEmptyString(photo200, photo100, photo200);
    }

    public boolean isGroupChannel() {
        return isGroupChannel;
    }

    public Conversation setGroupChannel(boolean isGroupChannel) {
        this.isGroupChannel = isGroupChannel;
        return this;
    }

    public int getMajor_id() {
        return major_id;
    }

    public Conversation setMajor_id(int major_id) {
        this.major_id = major_id;
        return this;
    }

    public int getMinor_id() {
        return minor_id;
    }

    public Conversation setMinor_id(int minor_id) {
        this.minor_id = minor_id;
        return this;
    }

    public static final class AclFlags {
        public static final int CAN_INVITE = 1;
        public static final int CAN_CHANGE_INFO = 2;
        public static final int CAN_CHANGE_PIN = 4;
        public static final int CAN_PROMOTE_USERS = 8;
        public static final int CAN_SEE_INVITE_LINK = 16;
        public static final int CAN_CHANGE_INVITE_LINK = 32;
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/model/Conversation.kt
package dev.ragnarok.fenrir.model

import android.content.Context
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.nonNullNoEmpty
import dev.ragnarok.fenrir.settings.Settings.get
import dev.ragnarok.fenrir.util.Utils.firstNonEmptyString

class Conversation(val id: Long) {
    var title: String? = null
        private set
    var unreadCount = 0
        private set
    var photo50: String? = null
        private set
    var photo100: String? = null
        private set
    var photo200: String? = null
        private set

    /**
     * ID of the last read incoming message.
     */
    var inRead = 0
        private set

    /**
     * ID of the last read outcoming message.
     */
    var outRead = 0
        private set
    var interlocutor: Owner? = null
        private set
    var currentKeyboard: Keyboard? = null
        private set
    var pinned: Message? = null
        private set
    var isGroupChannel = false
        private set
    var acl = 0
        private set
    var major_id = 0
        private set
    var minor_id = 0
        private set

    fun setAcl(acl: Int): Conversation {
        this.acl = acl
        return this
    }

    fun setPinned(pinned: Message?): Conversation {
        this.pinned = pinned
        return this
    }

    fun setInterlocutor(interlocutor: Owner?): Conversation {
        this.interlocutor = interlocutor
        return this
    }

    val imageUrl: String?
        get() {
            if (Peer.getType(id) == PeerType.CHAT || Peer.getType(id) == PeerType.CONTACT) {

                //if (isEmpty(img) && interlocutor != null) {
                //img = interlocutor.getMaxSquareAvatar();
                // }
                return firstNonEmptyString(photo200, photo100, photo50)
            }
            return interlocutor?.maxSquareAvatar
        }

    fun getDisplayTitle(context: Context): String? {
        return when (Peer.getType(id)) {
            PeerType.USER, PeerType.GROUP -> {
                val custom = get().main().getUserNameChanges(id)
                if (custom.nonNullNoEmpty()) {
                    return custom
                }
                if (interlocutor == null && title.isNullOrEmpty()) context.getString(R.string.unknown_first_name) + " " + context.getString(
                    R.string.unknown_last_name
                )
                else if (title.nonNullNoEmpty()) {
                    title
                } else interlocutor?.fullName
            }

            PeerType.CHAT, PeerType.CONTACT -> title
            else -> throw IllegalStateException("Unknown peer id: $id")
        }
    }

    fun getDisplayTitle(): String? {
        return when (Peer.getType(id)) {
            PeerType.USER, PeerType.GROUP -> {
                val custom = get().main().getUserNameChanges(id)
                if (custom.nonNullNoEmpty()) {
                    return custom
                }
                if (interlocutor == null && title.isNullOrEmpty()) null
                else if (title.nonNullNoEmpty()) {
                    title
                } else interlocutor?.fullName
            }

            PeerType.CHAT, PeerType.CONTACT -> title
            else -> throw IllegalStateException("Unknown peer id: $id")
        }
    }

    fun setTitle(title: String?): Conversation {
        this.title = title
        return this
    }

    fun setCurrentKeyboard(currentKeyboard: Keyboard?): Conversation {
        this.currentKeyboard = currentKeyboard
        return this
    }

    fun setUnreadCount(unreadCount: Int): Conversation {
        this.unreadCount = unreadCount
        return this
    }

    fun setPhoto50(photo50: String?): Conversation {
        this.photo50 = photo50
        return this
    }

    fun setPhoto100(photo100: String?): Conversation {
        this.photo100 = photo100
        return this
    }

    fun setPhoto200(photo200: String?): Conversation {
        this.photo200 = photo200
        return this
    }

    fun setInRead(inRead: Int): Conversation {
        this.inRead = inRead
        return this
    }

    fun setOutRead(outRead: Int): Conversation {
        this.outRead = outRead
        return this
    }

    fun setGroupChannel(isGroupChannel: Boolean): Conversation {
        this.isGroupChannel = isGroupChannel
        return this
    }

    fun setMajor_id(major_id: Int): Conversation {
        this.major_id = major_id
        return this
    }

    fun setMinor_id(minor_id: Int): Conversation {
        this.minor_id = minor_id
        return this
    }

    object AclFlags {
        const val CAN_INVITE = 1
        const val CAN_CHANGE_INFO = 2
        const val CAN_CHANGE_PIN = 4
        const val CAN_PROMOTE_USERS = 8
        const val CAN_SEE_INVITE_LINK = 16
        const val CAN_CHANGE_INVITE_LINK = 32
    }
}